Throttle case iteration for slower queries when running update rules

gherceg opened the pull request:
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->

## Technical Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->


### Migrations
<!-- Delete this section if the PR does not contain any migrations -->
<!-- https://commcare-hq.readthedocs.io/migrations_in_practice.html -->
- [ ] The migrations in this code can be safely applied first independently of the code

<!-- Please link to any past code changes that are coordinated with this migration -->

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [ ] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [ ] Risk label is set correctly
- [ ] The set of people pinged as reviewers is appropriate for the level of risk of the change

gherceg committed:
Throttle case iteration for slower queries when running update rules

dannyroberts commented on corehq/apps/data_interfaces/utils.py:
```diff
@@ -214,6 +215,23 @@ def _get_sql_repeat_record(domain, record_id):
 
+def _throttle_case_iterator(case_iterator):
+    """
+    Throttles slower queries that hog pgshard resources by sleeping for the amount of time the query took. Used
```
I might reword this. We're throttling when a query is slow, but just the fact that a query is slow doesn't necessarily mean it's hogging pgshard resources—in the pattern motivating this, any individual query's time is mostly spent _waiting_, not spent hogging CPU cycles or something.

So it's more that we want this entire workload (case rule processing) to gently back off when query times are high.

gherceg replied:
Thanks for clarifying. Updated in 6488628ba0a41a95a689a160da385adf6f2d706b.

dannyroberts commented on corehq/apps/data_interfaces/utils.py:
```diff
@@ -214,6 +215,23 @@ def _get_sql_repeat_record(domain, record_id):
+        end_time = time.time()
+        query_time = end_time - start_time
+        if query_time > 0.08:
+            time.sleep(query_time)
```
I think I still like
```suggestion
            time.sleep(query_time - 0.08)
```
or
```suggestion
            time.sleep(2 * (query_time - 0.08))
```

better, because there's no discontinuity in wait times. The way you have it here, wait time is at 80ms is 0ms and at 81 it's 81ms—big jump.

gherceg replied:
Great point. Updated in 6488628ba0a41a95a689a160da385adf6f2d706b

gherceg committed:
Address feedback

- Update formula for time to sleep to gently backoff rather than
abruptly go from 0 to 0.81 seconds of waiting.
- Update comment on throttle method to accurately reflect reasoning
behind this change
